# Data Cleaning 

The following code combines 1990-1999 songs and 2000-2009 songs from the Kaggle dataset and the Billboard Top Songs dataset, assigning label 1 to the Billboard Top Songs and 0 otherwise.

In [1]:

import pandas as pd 
import re

## Top 100 Songs Dataset

In [3]:

top_100= pd.read_csv('top_100_final.csv')
top_100

Year                 Song Title                          Artist  Label  \
0     1960  Theme from A Summer Place                     Percy Faith      1   
1     1960           He'll Have to Go                      Jim Reeves      1   
2     1960              Cathy's Clown             The Everly Brothers      1   
3     1960               Running Bear                  Johnny Preston      1   
4     1960                 Teen Angel                    Mark Dinning      1   
...    ...                        ...                             ...    ...   
6221  2022               Flower Shops  Ernest featuring Morgan Wallen      1   
6222  2022                To the Moon       Jnr Choi and Sam Tompkins      1   
6223  2022                     Unholy        Sam Smith and Kim Petras      1   
6224  2022            One Mississippi                      Kane Brown      1   
6225  2022   Circles Around This Town                    Maren Morris      1   

                                                 Lyrics  \
0     There's a summer placeWhere it may rain or sto...   
1     Put your sweet lips a little closer to the pho...   
2     Don't want your love anymoreDon't want your ki...   
3     *vocalizations*On the bank of the riverStood R...   
4     Teen AngelTeen AngelTeen AngelThat fateful nig...   
...                                                 ...   
6221  It's a beautiful day, she's been cryin' all ni...   
6222  Sit by myselfTalking to the moonTeh, ha, pull ...   
6223  Mummy don't know Daddy's getting hot At the Bo...   
6224  You and IHad this off and on so longYou've bee...   
6225  I swear I don't know how I trustedA Montero wi...   

                                                    URL           Genre  
0     https://genius.com/percy-faith-theme-from-a-su...  easy listening  
1     https://genius.com/jim-reeves-hell-have-to-go-...         country  
2     https://genius.com/the-everly-brothers-cathys-...            rock  
3     https://genius.com/johnny-preston-running-bear...             pop  
4     https://genius.com/mark-dinning-teen-angel-lyrics             pop  
...                                                 ...             ...  
6221      https://genius.com/ernest-flower-shops-lyrics             NaN  
6222     https://genius.com/jnr-choi-to-the-moon-lyrics             NaN  
6223                                                NaN             pop  
6224  https://genius.com/kane-brown-one-mississippi-...         country  
6225  https://genius.com/maren-morris-circles-around...         country  

[6226 rows x 7 columns]

## Labels for 1990-1999 songs

### Top 1990-1999 songs

In [4]:
top_90_99= top_100[(top_100['Year']>=1990 ) & (top_100['Year']<=1999)].reset_index(drop=True)
top_90_99

Year                                       Song Title  \
0    1990                                          Hold On   
1    1990                           It Must Have Been Love   
2    1990                             Nothing Compares 2 U   
3    1990                                           Poison   
4    1990                                            Vogue   
..    ...                                              ...   
992  1999           Better Days (And the Bottom Drops Out)   
993  1999                                Music of My Heart   
994  1999                                  Write This Down   
995  1999                                 When You Believe   
996  1999  (God Must Have Spent) A Little More Time on You   

                               Artist  Label  \
0                     Wilson Phillips      1   
1                             Roxette      1   
2                     Sinéad O'Connor      1   
3                      Bell Biv DeVoe      1   
4                             Madonna      1   
..                                ...    ...   
992                      Citizen King      1   
993          NSYNC and Gloria Estefan      1   
994                     George Strait      1   
995  Whitney Houston and Mariah Carey      1   
996                 Alabama and NSYNC      1   

                                                Lyrics  \
0    I know this pain Why do you lock yourself up i...   
1    Must have been loveBut it's over nowLay a whis...   
2    It's been seven hours and fifteen daysSince yo...   
3    Yeah, Spyderman and Freeze in full effectUh-hu...   
4    Strike a poseStrike a poseVogue Vogue Look aro...   
..                                                 ...   
992  In my shoes my toes are bustedMy kitchen says ...   
993  You'll never know what you've done for meWhat ...   
994  I never saw the end in sightFools are kind of ...   
995  Many nights we've prayedWith no proof anyone c...   
996  Can this be true? Tell me, can this be real?Ho...   

                                                   URL    Genre  
0    https://genius.com/wilson-phillips-hold-on-lyrics      pop  
1    https://genius.com/roxette-it-must-have-been-l...     rock  
2    https://genius.com/sinead-oconnor-nothing-comp...      pop  
3      https://genius.com/bell-biv-devoe-poison-lyrics       rb  
4              https://genius.com/madonna-vogue-lyrics      pop  
..                                                 ...      ...  
992  https://genius.com/citizen-king-better-days-an...    other  
993  https://genius.com/nsync-music-of-my-heart-lyrics      pop  
994  https://genius.com/george-strait-write-this-do...  country  
995  https://genius.com/whitney-houston-when-you-be...       rb  
996  https://genius.com/alabama-god-must-have-spent...      NaN  

[997 rows x 7 columns]

In [5]:
df_9= top_90_99.copy()
df_9

Year                                       Song Title  \
0    1990                                          Hold On   
1    1990                           It Must Have Been Love   
2    1990                             Nothing Compares 2 U   
3    1990                                           Poison   
4    1990                                            Vogue   
..    ...                                              ...   
992  1999           Better Days (And the Bottom Drops Out)   
993  1999                                Music of My Heart   
994  1999                                  Write This Down   
995  1999                                 When You Believe   
996  1999  (God Must Have Spent) A Little More Time on You   

                               Artist  Label  \
0                     Wilson Phillips      1   
1                             Roxette      1   
2                     Sinéad O'Connor      1   
3                      Bell Biv DeVoe      1   
4                             Madonna      1   
..                                ...    ...   
992                      Citizen King      1   
993          NSYNC and Gloria Estefan      1   
994                     George Strait      1   
995  Whitney Houston and Mariah Carey      1   
996                 Alabama and NSYNC      1   

                                                Lyrics  \
0    I know this pain Why do you lock yourself up i...   
1    Must have been loveBut it's over nowLay a whis...   
2    It's been seven hours and fifteen daysSince yo...   
3    Yeah, Spyderman and Freeze in full effectUh-hu...   
4    Strike a poseStrike a poseVogue Vogue Look aro...   
..                                                 ...   
992  In my shoes my toes are bustedMy kitchen says ...   
993  You'll never know what you've done for meWhat ...   
994  I never saw the end in sightFools are kind of ...   
995  Many nights we've prayedWith no proof anyone c...   
996  Can this be true? Tell me, can this be real?Ho...   

                                                   URL    Genre  
0    https://genius.com/wilson-phillips-hold-on-lyrics      pop  
1    https://genius.com/roxette-it-must-have-been-l...     rock  
2    https://genius.com/sinead-oconnor-nothing-comp...      pop  
3      https://genius.com/bell-biv-devoe-poison-lyrics       rb  
4              https://genius.com/madonna-vogue-lyrics      pop  
..                                                 ...      ...  
992  https://genius.com/citizen-king-better-days-an...    other  
993  https://genius.com/nsync-music-of-my-heart-lyrics      pop  
994  https://genius.com/george-strait-write-this-do...  country  
995  https://genius.com/whitney-houston-when-you-be...       rb  
996  https://genius.com/alabama-god-must-have-spent...      NaN  

[997 rows x 7 columns]

### Titles and Artists Cleaning

This function standardizes the input text by converting it to lowercase, removing unnecessary characters and words and ensuring consistent spacing. 
It is applied to standardize titles and artist names between the Kaggle dataset and the Billboard Top Songs dataset since they are extracted from two different sources (i.e., artist collaborations might be reported differently, using "and", "featuring", "feat", etc.). It ensures the correct combination of the two datasets.

In [6]:

def clean_text(text):
    #Convert text to lowercase and remove any character that is not alphanumeric or whitespace
    c_text = re.sub(r'[^a-zA-Z0-9\s]', '', text.lower())
    
    #Remove 'and', 'featuring', 'feat', and 'feat.' using word boundaries to match whole words
    c_text2 = re.sub(r'\b(and|featuring|feat\.?|feat\b)\b', '', c_text)
    
    #Replace multiple spaces with a single space and strip leading spaces
    c_text3 = re.sub(r'\s+', ' ', c_text2).strip()
    
    return c_text3
    

In [7]:
titles= top_90_99['Song Title'].tolist()
titles= [clean_text(title) for title in titles]
print(titles)

authors= top_90_99['Artist'].tolist()
authors= [clean_text(author) for author in authors]
print(authors)



artist_songs_dict = {}
for author, title in zip(authors, titles):
    if author in artist_songs_dict:
        artist_songs_dict[author].append(title)
    else:
        artist_songs_dict[author] = [title]


print(artist_songs_dict)

['hold on', 'it must have been love', 'nothing compares 2 u', 'poison', 'vogue', 'vision of love', 'another day in paradise', 'hold on', 'cradle of love', 'blaze of glory', 'do me', 'how am i supposed to live without you', 'pump up the jam', 'opposites attract', 'escapade', 'all i wanna do is make love to you', 'close to you', 'black velvet', 'release me', 'dont know much', 'all around the world', 'i wanna be rich', 'rub you the right way', 'she aint worth it', 'if wishes came true', 'the power', 'cant live without your love affection', 'love will lead you back', 'dont wanna fall in love', 'two to make it right', 'sending all my love', 'unskinny bop', 'step by step', 'dangerous', 'we didnt start the fire', 'i dont have the heart', 'downtown train', 'rhythm nation', 'ill be your everything', 'roam', 'everything', 'back to life however do you want me', 'here now', 'alright', 'ice ice baby', 'blame it on the rain', 'have you seen her', 'with every beat of my heart', 'come back to me', 'no

After standardizing the artist names, we create a dictionary where the key is the artist's name and the value is a dictionary containing the associated songs.


### Kaggle Dataset Songs

In [8]:
data_90_99_1= pd.read_parquet('1990_1999_part1.parquet')
data_90_99_2= pd.read_parquet('1990_1999_part2.parquet')

data_90_99= pd.concat([data_90_99_1, data_90_99_2])
data_90_99.reset_index(drop=True)

title   tag                  artist  year  \
0                        Can I Live   rap                   JAY-Z  1996   
1                 Rockin and Rollin   rap                 Cam'ron  1998   
2                            DEvils   rap                   JAY-Z  1996   
3          Its Hot Some Like It Hot   rap                   JAY-Z  1999   
4                     Its Like That   rap                   JAY-Z  1998   
...                             ...   ...                     ...   ...   
222922                     Since 92   pop  Undeclinable Ambuscade  1998   
222923  Horror Story Latnem Version   rap        House Of Krazees  1994   
222924        From Season to Season   pop          Louis Philippe  1990   
222925           Long Way From Love  rock             Marcie Free  1993   
222926               Blue September   pop          Windows (Jazz)  1990   

         views                                             lyrics  
0       468624  "   Yeah, hah, yeah, Roc-A-Fella We invite you...  
1         6399  " Ay yo you wonder who I are I guzzle up at th...  
2       504959                          "   ""Dear God – I wonder  
3       103549  "   Yo, show closer, J-to-the-A-Y-Hovah Place ...  
4        37692  " Yeah, uh-huh, watch this y'all Uhh, watch th...  
...        ...                                                ...  
222922       2  We're here since '92 If that's not punk enough...  
222923       4   Next motherfuckers's gonna get my metal   I c...  
222924       4  When there's nothing else I can do Except thin...  
222925       3  "Is it tenderness or torture Joy that only end...  
222926       2   Looking out across the empty streets Wonderin...  

[222927 rows x 6 columns]

In [9]:
#Initially, we set all the labels to 0
data_90_99['Label']=0


In [10]:
#This function cleans the song titles

def clean_song(song): 
    clean_song= re.sub(r'"', '', song)
    return clean_song



In [11]:
data_90_99['lyrics']= data_90_99['lyrics'].apply(clean_song)
data_90_99= data_90_99.reset_index(drop=True)

In [12]:
data_90_99

title   tag                  artist  year  \
0                        Can I Live   rap                   JAY-Z  1996   
1                 Rockin and Rollin   rap                 Cam'ron  1998   
2                            DEvils   rap                   JAY-Z  1996   
3          Its Hot Some Like It Hot   rap                   JAY-Z  1999   
4                     Its Like That   rap                   JAY-Z  1998   
...                             ...   ...                     ...   ...   
222922                     Since 92   pop  Undeclinable Ambuscade  1998   
222923  Horror Story Latnem Version   rap        House Of Krazees  1994   
222924        From Season to Season   pop          Louis Philippe  1990   
222925           Long Way From Love  rock             Marcie Free  1993   
222926               Blue September   pop          Windows (Jazz)  1990   

         views                                             lyrics  Label  
0       468624     Yeah, hah, yeah, Roc-A-Fella We invite you ...      0  
1         6399   Ay yo you wonder who I are I guzzle up at the...      0  
2       504959                                Dear God – I wonder      0  
3       103549     Yo, show closer, J-to-the-A-Y-Hovah Place s...      0  
4        37692   Yeah, uh-huh, watch this y'all Uhh, watch thi...      0  
...        ...                                                ...    ...  
222922       2  We're here since '92 If that's not punk enough...      0  
222923       4   Next motherfuckers's gonna get my metal   I c...      0  
222924       4  When there's nothing else I can do Except thin...      0  
222925       3  Is it tenderness or torture Joy that only ends...      0  
222926       2   Looking out across the empty streets Wonderin...      0  

[222927 rows x 7 columns]

We update the 'Label' column in the Kaggle Dataset to 1 if a "Billboard Top Song" is already in the dataset. It removes matching entries from the "Billboard Top Songs" dataset to check if the combination of top songs in the two datasets sum up to 997 (they are not exactly equal to 1000 since we removed some songs without lyrics).

In [13]:

for index, row in data_90_99.iterrows():
    artist_cleaned = clean_text(row['artist'])
    title_cleaned = clean_text(row['title'])
    
    if artist_cleaned in artist_songs_dict and title_cleaned in artist_songs_dict[artist_cleaned]:
        data_90_99.loc[index, 'Label'] = 1

       
        condition = (top_90_99['Artist'].apply(clean_text) == artist_cleaned) & (top_90_99['Song Title'].apply(clean_text) == title_cleaned)
        top_90_99 = top_90_99.drop(top_90_99[condition].index).reset_index(drop=True)

In [14]:
#Check how many top songs are already in the dataset
data_90_99[data_90_99['Label']==1]

title  tag                artist  year  \
21                   1st of tha Month  rap  Bone Thugs-N-Harmony  1995   
36                  Elevators Me  You  rap               OutKast  1996   
41                             I Wish  rap               Skee-Lo  1995   
50      Hard Knock Life Ghetto Anthem  rap                 JAY-Z  1998   
69                  It Was a Good Day  rap              Ice Cube  1992   
...                               ...  ...                   ...   ...   
198791                             If   rb         Janet Jackson  1993   
198794                Because of Love   rb         Janet Jackson  1993   
198894       You Mean The World to Me   rb          Toni Braxton  1994   
200696      Dont Cry for Me Argentina  pop               Madonna  1997   
213678                Go the Distance  pop        Michael Bolton  1997   

          views                                             lyrics  Label  
21       266526   Wake up, wake up, wake up, wake up, wake up I...      1  
36       325019   One for the money, yes, sir, two for the show...      1  
41       459740   Hey, this is radio station WSKEE We're taking...      1  
50       489194     Take the bass line out Uh-huh, Jigga Uh-huh...      1  
69      1463692     Break 'em Yeah Yeah Yeah Uh   Just wakin' u...      1  
...         ...                                                ...    ...  
198791    88343   Sittin' over here, starin' in your face With ...      1  
198794     3761   You got me singing Shoo-doo-doo Shoo-shoo-doo...      1  
198894    18124   If you could give me one good reason Why I sh...      1  
200696    23256   It won't be easy, you'll think it's strange W...      1  
213678      809  I have often dreamed, of a far off place Where...      1  

[698 rows x 7 columns]

In [15]:
top_90_99

Year                                       Song Title  \
0    1990                          Another Day in Paradise   
1    1990            How Am I Supposed to Live Without You   
2    1990                                  Pump Up the Jam   
3    1990                                Opposites Attract   
4    1990                                         Escapade   
..    ...                                              ...   
248  1999                                             Iris   
249  1999                                      Satisfy You   
250  1999                                Music of My Heart   
251  1999                                 When You Believe   
252  1999  (God Must Have Spent) A Little More Time on You   

                               Artist  Label  \
0                        Phil Collins      1   
1                      Michael Bolton      1   
2                        Technotronic      1   
3                         Paula Abdul      1   
4                       Janet Jackson      1   
..                                ...    ...   
248                     Goo Goo Dolls      1   
249     Puff Daddy featuring R. Kelly      1   
250          NSYNC and Gloria Estefan      1   
251  Whitney Houston and Mariah Carey      1   
252                 Alabama and NSYNC      1   

                                                Lyrics  \
0    She calls out to the man on the street"Sir, ca...   
1    I could hardly believe it, when I heard the ne...   
2    Pump up the jamPump it upWhile your feet are s...   
3    Baby, it seems we never ever agreeYou like the...   
4    As I was walkin' bySaw you standin' there with...   
..                                                 ...   
248  And I'd give up forever to touch you'Cause I k...   
249  All I want is somebody who's gonna love me for...   
250  You'll never know what you've done for meWhat ...   
251  Many nights we've prayedWith no proof anyone c...   
252  Can this be true? Tell me, can this be real?Ho...   

                                                   URL           Genre  
0    https://genius.com/phil-collins-another-day-in...            rock  
1    https://genius.com/michael-bolton-how-am-i-sup...  easy listening  
2    https://genius.com/technotronic-pump-up-the-ja...             pop  
3    https://genius.com/paula-abdul-opposites-attra...             pop  
4     https://genius.com/janet-jackson-escapade-lyrics             pop  
..                                                 ...             ...  
248       https://genius.com/goo-goo-dolls-iris-lyrics            rock  
249   https://genius.com/puff-daddy-satisfy-you-lyrics         hip hop  
250  https://genius.com/nsync-music-of-my-heart-lyrics             pop  
251  https://genius.com/whitney-houston-when-you-be...              rb  
252  https://genius.com/alabama-god-must-have-spent...             NaN  

[253 rows x 7 columns]

In [20]:
len(data_90_99[data_90_99['Label']==1])+len(top_90_99)

951

The combination of top songs are not exactly equal to 997 because there could be some duplicates which have been dropped.

In [16]:
df1 = df_9[['Artist', 'Song Title']].map(clean_text)
df2= data_90_99[['artist', 'title']].map(clean_text).rename(columns={'artist': 'Artist', 'title': 'Song Title'})
merged_df = pd.merge(df1, df2, on=['Artist', 'Song Title'], how='inner')
merged_df

Artist                        Song Title
0    wilson phillips                           hold on
1            roxette            it must have been love
2      sinad oconnor              nothing compares 2 u
3     bell biv devoe                            poison
4            madonna                             vogue
..               ...                               ...
739         fastball                    out of my head
740             jayz     hard knock life ghetto anthem
741    joey mcintyre                     stay the same
742     citizen king  better days the bottom drops out
743    george strait                   write this down

[744 rows x 2 columns]

In [17]:
merged_df[merged_df.duplicated()].value_counts().sum()


46

Indeed, there are 46 duplicates and removing them we have a total of 698 top songs already in the Kaggle Dataset.

### Concatenation of the two datasets

Finally, we concatenate the previously obtained dataset with the remaining top songs that were not already present in the Kaggle dataset.

In [167]:

titles= top_90_99['Song Title'].tolist()
tags= top_90_99['Genre'].tolist()
artists= top_90_99['Artist'].tolist()
years= top_90_99['Year'].tolist()
lyrics= top_90_99['Lyrics'].tolist()
label= top_90_99['Label'].tolist()

top_90_99= pd.DataFrame({'title': titles, 'tag': tags, 'artist': artists, 'year': years, 'lyrics': lyrics, 'Label': label})


data_90_99= data_90_99.drop(columns=['views'])


total90_99= pd.concat([data_90_99, top_90_99]).reset_index(drop=True)
total90_99.to_csv('total90_99.csv', index= False)


In [168]:
total90_99

title      tag  \
0                                            Can I Live      rap   
1                                     Rockin and Rollin      rap   
2                                                DEvils      rap   
3                              Its Hot Some Like It Hot      rap   
4                                         Its Like That      rap   
...                                                 ...      ...   
223175                                             Iris     rock   
223176                                      Satisfy You  hip hop   
223177                                Music of My Heart      pop   
223178                                 When You Believe       rb   
223179  (God Must Have Spent) A Little More Time on You      NaN   

                                  artist  year  \
0                                  JAY-Z  1996   
1                                Cam'ron  1998   
2                                  JAY-Z  1996   
3                                  JAY-Z  1999   
4                                  JAY-Z  1998   
...                                  ...   ...   
223175                     Goo Goo Dolls  1999   
223176     Puff Daddy featuring R. Kelly  1999   
223177          NSYNC and Gloria Estefan  1999   
223178  Whitney Houston and Mariah Carey  1999   
223179                 Alabama and NSYNC  1999   

                                                   lyrics  Label  
0          Yeah, hah, yeah, Roc-A-Fella We invite you ...      0  
1        Ay yo you wonder who I are I guzzle up at the...      0  
2                                     Dear God – I wonder      0  
3          Yo, show closer, J-to-the-A-Y-Hovah Place s...      0  
4        Yeah, uh-huh, watch this y'all Uhh, watch thi...      0  
...                                                   ...    ...  
223175  And I'd give up forever to touch you'Cause I k...      1  
223176  All I want is somebody who's gonna love me for...      1  
223177  You'll never know what you've done for meWhat ...      1  
223178  Many nights we've prayedWith no proof anyone c...      1  
223179  Can this be true? Tell me, can this be real?Ho...      1  

[223180 rows x 6 columns]

#### We repeat the exact same procedure for 2000-2009 songs

## Labels for 2000-2009 songs

### Top 2000-2009 songs

In [24]:
top_00_09= top_100[(top_100['Year']>=2000 ) & (top_100['Year']<=2009)].reset_index(drop=True)
top_00_09

Year            Song Title  \
0    2000               Breathe   
1    2000                Smooth   
2    2000           Maria Maria   
3    2000          I Wanna Know   
4    2000   Everything You Want   
..    ...                   ...   
994  2009               Goodbye   
995  2009  Say Hey (I Love You)   
996  2009         Pop Champagne   
997  2009          Pretty Wings   
998  2009       Never Say Never   

                                                Artist  Label  \
0                                           Faith Hill      1   
1                         Santana featuring Rob Thomas      1   
2                    Santana featuring The Product G&B      1   
3                                                  Joe      1   
4                                     Vertical Horizon      1   
..                                                 ...    ...   
994                                  Kristinia DeBarge      1   
995  Michael Franti & Spearhead featuring Cherine A...      1   
996    Jim Jones featuring Ron Browz and Juelz Santana      1   
997                                            Maxwell      1   
998                                           The Fray      1   

                                                Lyrics  \
0    I can feel the magic floating in the airBeing ...   
1    Man, it's a hot oneLike seven inches from the ...   
2    Ladies and gents, turn up your sound systemsTo...   
3    YeahOh-oh, yeahAlright, oh, oh, ohIt's amazing...   
4    Somewhere there's speakingIt's already coming ...   
..                                                 ...   
994  Am I supposed to put my life on holdBecause yo...   
995  Comme da selecta Ayy, uh-huh, woo (That's righ...   
996  Ether boy! Hey!How we ball in the club, I know...   
997  Time will bring the real end of our trialOne d...   
998  Some things we don't talk aboutRather do witho...   

                                                   URL    Genre  
0         https://genius.com/faith-hill-breathe-lyrics      pop  
1             https://genius.com/santana-smooth-lyrics     rock  
2        https://genius.com/santana-maria-maria-lyrics      NaN  
3           https://genius.com/joe-i-wanna-know-lyrics       rb  
4    https://genius.com/vertical-horizon-everything...     rock  
..                                                 ...      ...  
994  https://genius.com/kristinia-debarge-goodbye-l...      pop  
995                                                NaN  hip hop  
996  https://genius.com/jim-jones-pop-champagne-lyrics      NaN  
997     https://genius.com/maxwell-pretty-wings-lyrics     soul  
998  https://genius.com/the-fray-never-say-never-ly...     rock  

[999 rows x 7 columns]

In [25]:
df_10= top_00_09.copy()
df_10

Year            Song Title  \
0    2000               Breathe   
1    2000                Smooth   
2    2000           Maria Maria   
3    2000          I Wanna Know   
4    2000   Everything You Want   
..    ...                   ...   
994  2009               Goodbye   
995  2009  Say Hey (I Love You)   
996  2009         Pop Champagne   
997  2009          Pretty Wings   
998  2009       Never Say Never   

                                                Artist  Label  \
0                                           Faith Hill      1   
1                         Santana featuring Rob Thomas      1   
2                    Santana featuring The Product G&B      1   
3                                                  Joe      1   
4                                     Vertical Horizon      1   
..                                                 ...    ...   
994                                  Kristinia DeBarge      1   
995  Michael Franti & Spearhead featuring Cherine A...      1   
996    Jim Jones featuring Ron Browz and Juelz Santana      1   
997                                            Maxwell      1   
998                                           The Fray      1   

                                                Lyrics  \
0    I can feel the magic floating in the airBeing ...   
1    Man, it's a hot oneLike seven inches from the ...   
2    Ladies and gents, turn up your sound systemsTo...   
3    YeahOh-oh, yeahAlright, oh, oh, ohIt's amazing...   
4    Somewhere there's speakingIt's already coming ...   
..                                                 ...   
994  Am I supposed to put my life on holdBecause yo...   
995  Comme da selecta Ayy, uh-huh, woo (That's righ...   
996  Ether boy! Hey!How we ball in the club, I know...   
997  Time will bring the real end of our trialOne d...   
998  Some things we don't talk aboutRather do witho...   

                                                   URL    Genre  
0         https://genius.com/faith-hill-breathe-lyrics      pop  
1             https://genius.com/santana-smooth-lyrics     rock  
2        https://genius.com/santana-maria-maria-lyrics      NaN  
3           https://genius.com/joe-i-wanna-know-lyrics       rb  
4    https://genius.com/vertical-horizon-everything...     rock  
..                                                 ...      ...  
994  https://genius.com/kristinia-debarge-goodbye-l...      pop  
995                                                NaN  hip hop  
996  https://genius.com/jim-jones-pop-champagne-lyrics      NaN  
997     https://genius.com/maxwell-pretty-wings-lyrics     soul  
998  https://genius.com/the-fray-never-say-never-ly...     rock  

[999 rows x 7 columns]

### Titles and Artists Cleaning

This function standardizes the input text by converting it to lowercase, removing unnecessary characters and words and ensuring consistent spacing. 
It is applied to standardize titles and artist names between the Kaggle dataset and the Billboard Top Songs dataset since they are extracted from two different sources (i.e., artist collaborations might be reported differently, using "and", "featuring", "feat", etc.). This ensures the correct combination of the two datasets.

In [26]:

def clean_text(text):
    #Convert text to lowercase and remove any character that is not alphanumeric or whitespace
    c_text = re.sub(r'[^a-zA-Z0-9\s]', '', text.lower())
    
    #Remove 'and', 'featuring', 'feat', and 'feat.' using word boundaries to match whole words
    c_text2 = re.sub(r'\b(and|featuring|feat\.?|feat\b)\b', '', c_text)
    
    #Replace multiple spaces with a single space and strip leading spaces
    c_text3 = re.sub(r'\s+', ' ', c_text2).strip()
    
    return c_text3

In [27]:

titles= top_00_09['Song Title'].tolist()
titles= [clean_text(title) for title in titles]
print(titles)

authors= top_00_09['Artist'].tolist()
authors= [clean_text(author) for author in authors]
print(authors)



artist_songs_dict = {}
for author, title in zip(authors, titles):
    if author in artist_songs_dict:
        artist_songs_dict[author].append(title)
    else:
        artist_songs_dict[author] = [title]


print(artist_songs_dict)

['breathe', 'smooth', 'maria maria', 'i wanna know', 'everything you want', 'say my name', 'i knew i loved you', 'amazed', 'bent', 'he wasnt man enough', 'higher', 'try again', 'jumpin jumpin', 'thong song', 'kryptonite', 'there you go', 'music', 'doesnt really matter', 'what a girl wants', 'back at one', 'bye bye bye', 'you sang to me', 'i need to know', 'get it on tonite', 'incomplete', 'i try', 'its gonna be me', 'thats the way it is', 'country grammar hot shit', 'bring it all to me', 'show me the meaning of being lonely', 'hot boyz', 'back here', 'it feels so good', 'absolutely story of a girl', 'with arms wide open', 'be with you', 'come on over baby all i want is you', 'no more', 'all the small things', 'the way you love me', 'i turn to you', 'never let you go', 'i need you', 'thank god i found you', 'lets get married', 'my love is your love', 'then the morning comes', 'blue da ba dee', 'desert rose', 'the real slim shady', 'most girls', 'wifey', 'wonderful', 'oops i did it again

After standardizing the artist names, we create a dictionary where the key is the artist's name and the value is a dictionary containing the associated songs.

### Kaggle Dataset Songs

In [28]:
data_00_09_1= pd.read_parquet('2000_2009_part1.parquet')
data_90_99_2= pd.read_parquet('2000_2009_part2.parquet')

data_00_09= pd.concat([data_00_09_1, data_90_99_2])
data_00_09


title   tag              artist  year   views  \
0                      Killa Cam   rap             Cam'ron  2004  173166   
1              Forgive Me Father   rap            Fabolous  2003    4743   
2                   Down and Out   rap             Cam'ron  2004  144404   
3                         Fly In   rap           Lil Wayne  2005   78271   
4                 Lollipop Remix   rap           Lil Wayne  2008  580832   
...                          ...   ...                 ...   ...     ...   
218784         Opposites Detract  rock          Dyscarnate  2008       1   
218785  Kid Youll Move Mountains  rock  Junior Astronomers  2009       5   
218786        Attacked by Whales  rock  Junior Astronomers  2009       3   
218787  Societys Child Live 2001   pop           Janis Ian  2001       4   
218788      Im Going to Kill You  rock       Wesley Willis  2001       3   

                                                   lyrics  
0       " Killa Cam, Killa Cam, Cam Killa Cam, Killa C...  
1       "Maybe cause I'm eatin And these bastards fien...  
2       "   Ugh, Killa! Baby! Kanye, this that 1970s H...  
3       " So they ask me ""Young boy What you gon' do ...  
4       " Haha Uh-huh No homo (Young Mula, baby!) I sa...  
...                                                   ...  
218784  Crimson faced and riddled with scars. Caricatu...  
218785  "(Verse 1) They say, ""Focus!"" All you are is...  
218786  She asked my opinion I told her the truth You'...  
218787  " You come to my door, baby Face is clean and ...  
218788  " Look out, stupid jerk I'm going to break int...  

[452203 rows x 6 columns]

In [29]:
#Initially, we set all the labels to 0
data_00_09['Label']=0


In [30]:
#This function cleans the song titles
def clean_song(song): 
    clean_song= re.sub(r'"', '', song)
    return clean_song

In [31]:
data_00_09['lyrics']= data_00_09['lyrics'].apply(clean_song)
data_00_09.reset_index(drop=True)

title   tag              artist  year   views  \
0                      Killa Cam   rap             Cam'ron  2004  173166   
1              Forgive Me Father   rap            Fabolous  2003    4743   
2                   Down and Out   rap             Cam'ron  2004  144404   
3                         Fly In   rap           Lil Wayne  2005   78271   
4                 Lollipop Remix   rap           Lil Wayne  2008  580832   
...                          ...   ...                 ...   ...     ...   
452198         Opposites Detract  rock          Dyscarnate  2008       1   
452199  Kid Youll Move Mountains  rock  Junior Astronomers  2009       5   
452200        Attacked by Whales  rock  Junior Astronomers  2009       3   
452201  Societys Child Live 2001   pop           Janis Ian  2001       4   
452202      Im Going to Kill You  rock       Wesley Willis  2001       3   

                                                   lyrics  Label  
0        Killa Cam, Killa Cam, Cam Killa Cam, Killa Ca...      0  
1       Maybe cause I'm eatin And these bastards fiend...      0  
2          Ugh, Killa! Baby! Kanye, this that 1970s He...      0  
3        So they ask me Young boy What you gon' do the...      0  
4        Haha Uh-huh No homo (Young Mula, baby!) I say...      0  
...                                                   ...    ...  
452198  Crimson faced and riddled with scars. Caricatu...      0  
452199  (Verse 1) They say, Focus! All you are is not ...      0  
452200  She asked my opinion I told her the truth You'...      0  
452201   You come to my door, baby Face is clean and s...      0  
452202   Look out, stupid jerk I'm going to break into...      0  

[452203 rows x 7 columns]

In [32]:
data_00_09= data_00_09[data_00_09['title'].notnull()].reset_index(drop=True)

We update the 'Label' column in the Kaggle Dataset to 1 if a "Billboard Top Song" is already in the dataset. It removes matching entries from the "Billboard Top Songs" dataset to check if the combination of top songs in the two datasets sum up to 999 (they are not exactly equal to 1000 since we removed some songs without lyrics).

In [33]:

for index, row in data_00_09.iterrows():

    artist_cleaned = clean_text(row['artist'])
    title_cleaned = clean_text(row['title'])
    
    if artist_cleaned in artist_songs_dict and title_cleaned in artist_songs_dict[artist_cleaned]:
        data_00_09.loc[index, 'Label'] = 1

       
        condition = (top_00_09['Artist'].apply(clean_text) == artist_cleaned) & (top_00_09['Song Title'].apply(clean_text) == title_cleaned)
        top_00_09 = top_00_09.drop(top_00_09[condition].index).reset_index(drop=True)

In [34]:
#Check how many top songs are already in the dataset
data_00_09[data_00_09['Label']==1]

title      tag  \
48                           A Milli      rap   
59                       We Fly High      rap   
100                          Forever      rap   
139                    Lose Yourself      rap   
163                       In da Club      rap   
...                              ...      ...   
396397              I Hope You Dance  country   
405253            One Step at a Time      pop   
405330  It’s A Great Day To Be Alive  country   
405564                  Wait for You      pop   
446882                  Pretty Wings       rb   

                                       artist  year    views  \
48                                  Lil Wayne  2008  1237174   
59                                  Jim Jones  2006    63337   
100     Drake, Kanye West, Lil Wayne & Eminem  2009  2061952   
139                                    Eminem  2002  6263971   
163                                   50 Cent  2003  2005139   
...                                       ...   ...      ...   
396397                         Lee Ann Womack  2000    80415   
405253                          Jordin Sparks  2007    32536   
405330                           Travis Tritt  2000    28018   
405564                          Elliott Yamin  2007    44614   
446882                                Maxwell  2009      253   

                                                   lyrics  Label  
48       Bangladesh Young Money! You dig? Mack, I'm go...      1  
59       I wear a mean dark pair of shades And you can...      1  
100      It may not mean nothin' to y'all But understa...      1  
139      Look, if you had one shot or one opportunity ...      1  
163        Go, go, go, go, go, go Go shawty, it's your...      1  
...                                                   ...    ...  
396397   I hope you never lose your sense of wonder Yo...      1  
405253   Hurry up and wait So close, but so far away E...      1  
405330   I got rice cookin' in the microwave Got a thr...      1  
405564   I never felt nothing in the world like this b...      1  
446882  Time will bring the real end of our trial One ...      1  

[569 rows x 7 columns]

In [35]:
top_00_09

Year            Song Title  \
0    2000               Breathe   
1    2000                Smooth   
2    2000           Maria Maria   
3    2000   Everything You Want   
4    2000           Say My Name   
..    ...                   ...   
384  2009     How Do You Sleep?   
385  2009          I Run to You   
386  2009           Green Light   
387  2009  Say Hey (I Love You)   
388  2009         Pop Champagne   

                                                Artist  Label  \
0                                           Faith Hill      1   
1                         Santana featuring Rob Thomas      1   
2                    Santana featuring The Product G&B      1   
3                                     Vertical Horizon      1   
4                                      Destiny's Child      1   
..                                                 ...    ...   
384                 Jesse McCartney featuring Ludacris      1   
385                                    Lady Antebellum      1   
386                   John Legend featuring André 3000      1   
387  Michael Franti & Spearhead featuring Cherine A...      1   
388    Jim Jones featuring Ron Browz and Juelz Santana      1   

                                                Lyrics  \
0    I can feel the magic floating in the airBeing ...   
1    Man, it's a hot oneLike seven inches from the ...   
2    Ladies and gents, turn up your sound systemsTo...   
3    Somewhere there's speakingIt's already coming ...   
4    Say my name, say my nameIf no one is around yo...   
..                                                 ...   
384  Oh-oh, oh-oh, oh-oh, oh-ohOh-oh, oh-oh, oh-oh,...   
385  I run from hateI run from prejudiceI run from ...   
386  Give me the green lightGive me just one nightI...   
387  Comme da selecta Ayy, uh-huh, woo (That's righ...   
388  Ether boy! Hey!How we ball in the club, I know...   

                                                   URL    Genre  
0         https://genius.com/faith-hill-breathe-lyrics      pop  
1             https://genius.com/santana-smooth-lyrics     rock  
2        https://genius.com/santana-maria-maria-lyrics      NaN  
3    https://genius.com/vertical-horizon-everything...     rock  
4    https://genius.com/destinys-child-say-my-name-...       rb  
..                                                 ...      ...  
384  https://genius.com/jesse-mccartney-how-do-you-...      pop  
385  https://genius.com/lady-antebellum-i-run-to-yo...  country  
386  https://genius.com/john-legend-green-light-lyrics     soul  
387                                                NaN  hip hop  
388  https://genius.com/jim-jones-pop-champagne-lyrics      NaN  

[389 rows x 7 columns]

In [37]:
len(data_00_09[data_00_09['Label']==1])+len(top_00_09)

958

The combination of top songs are not exactly equal to 999 because there could be some duplicates which have been dropped.

In [38]:
df1 = df_10[['Artist', 'Song Title']].map(clean_text)
df2= data_00_09[['artist', 'title']].map(clean_text).rename(columns={'artist': 'Artist', 'title': 'Song Title'})
merged_df = pd.merge(df1, df2, on=['Artist', 'Song Title'], how='inner')
merged_df

Artist           Song Title
0                  joe         i wanna know
1      matchbox twenty                 bent
2         toni braxton  he wasnt man enough
3              aaliyah            try again
4         3 doors down           kryptonite
..                 ...                  ...
605     zac brown band       whatever it is
606     kelly clarkson         already gone
607  kristinia debarge              goodbye
608            maxwell         pretty wings
609           the fray      never say never

[610 rows x 2 columns]

In [39]:
merged_df[merged_df.duplicated()].value_counts().sum()

41

Indeed, there are 41 duplicates and removing them we have a total of 569 top songs already present in the Kaggle Dataset.

### Combination of the two datasets 

Finally, we concatenate the previously obtained dataset with the remaining top songs that were not already present in the Kaggle dataset.

In [185]:
titles= top_00_09['Song Title'].tolist()
tags= top_00_09['Genre'].tolist()
artists= top_00_09['Artist'].tolist()
years= top_00_09['Year'].tolist()
lyrics= top_00_09['Lyrics'].tolist()
label= top_00_09['Label'].tolist()

top_00_09= pd.DataFrame({'title': titles, 'tag': tags, 'artist': artists, 'year': years, 'lyrics': lyrics, 'Label': label})


data_00_09= data_00_09.drop(columns=['views'])


total00_09= pd.concat([data_00_09, top_00_09]).reset_index(drop=True)
total00_09.to_csv('total00_09.csv', index= False)


In [186]:
total00_09

title      tag  \
0                  Killa Cam      rap   
1          Forgive Me Father      rap   
2               Down and Out      rap   
3                     Fly In      rap   
4             Lollipop Remix      rap   
...                      ...      ...   
452582     How Do You Sleep?      pop   
452583          I Run to You  country   
452584           Green Light     soul   
452585  Say Hey (I Love You)  hip hop   
452586         Pop Champagne      NaN   

                                                   artist  year  \
0                                                 Cam'ron  2004   
1                                                Fabolous  2003   
2                                                 Cam'ron  2004   
3                                               Lil Wayne  2005   
4                                               Lil Wayne  2008   
...                                                   ...   ...   
452582                 Jesse McCartney featuring Ludacris  2009   
452583                                    Lady Antebellum  2009   
452584                   John Legend featuring André 3000  2009   
452585  Michael Franti & Spearhead featuring Cherine A...  2009   
452586    Jim Jones featuring Ron Browz and Juelz Santana  2009   

                                                   lyrics  Label  
0        Killa Cam, Killa Cam, Cam Killa Cam, Killa Ca...      0  
1       Maybe cause I'm eatin And these bastards fiend...      0  
2          Ugh, Killa! Baby! Kanye, this that 1970s He...      0  
3        So they ask me Young boy What you gon' do the...      0  
4        Haha Uh-huh No homo (Young Mula, baby!) I say...      0  
...                                                   ...    ...  
452582  Oh-oh, oh-oh, oh-oh, oh-ohOh-oh, oh-oh, oh-oh,...      1  
452583  I run from hateI run from prejudiceI run from ...      1  
452584  Give me the green lightGive me just one nightI...      1  
452585  Comme da selecta Ayy, uh-huh, woo (That's righ...      1  
452586  Ether boy! Hey!How we ball in the club, I know...      1  

[452587 rows x 6 columns]